Add pdf generation dependencies and configure lambda layer

julialeague opened the pull request:
From: https://www.notion.so/Deploy-the-wkhtmltopdf-tool-as-an-AWS-Lambda-layer-3db0df1b95a24e5fad9211af871eadd1

This PR introduces pdfkit and wkhtmltopdf as dependencies and adds configuration that allows for the deployment of wkhtmltopdf as a lambda layer. The reason we need to deploy wkhtmltopdf is because it's an OS command line tool rather than a Python library. pdfkit (and many other PDF generation libraries) depends on wkhtmltopdf. Wkhtmltopdf needed to be compiled for use in AmazonLinux2, and that compiled package is located in the `vro-main/layer` directory.

Code was added to the vro-main lambda to test that pdfkit could communicate with wkhtmltopdf. I added some basic tests for the added code also.

In addition to the above, I also added a `Stage` environment variable so that I could use that to toggle off the calls to other AWS services (in my case, the Secrets Manager) while I tested my additions locally. The 'dev' stage is used for that purpose right now. The 'prod' stage is the deployed environment.

You can test locally by running `make pytest` and `make local.sam.invoke`. Note that you need to install `wkhtmltopdf` locally now to run pytest. Also note that you should refresh your local cf-template-params.env file to include the Stage variable, and set it to your intended environment ('dev' or 'prod', currently).

I intend to log some follow-up cards to address some issues related to this work **Update - they're logged now!** :
* fix the broken deployment (move ALL dependencies into another layer) **I wasn't able to test the final deployment to AWS because of this blocker** : [card logged](https://www.notion.so/Move-all-dependencies-into-lambda-layer-c67d96d19d5b4dcdb0f00c600f25b505)
* determine whether we can reference the wkhtmltopdf Amazon Linux 2 build for our SAM template instead of hosting the binary in our repo : [card logged](https://www.notion.so/Update-wkhtmltopdf-binary-reference-in-SAM-template-9cc64f9882b7457d99c743cdb091ab22)
* establish an ADR process and update with wkhtmltopdf and pdfkit + lambda layers [card logged](https://www.notion.so/Establish-an-ADR-process-060287a0046d4cb3830f2ca19f98fcfb)

julialeague commented on hypertension/.vscode/launch.json:
```diff
@@ -1,5 +1,29 @@
+                }
+            }
+        },
         {
```
I added my local debug config for the VroMain function since this file was already checked in.

julialeague commented on hypertension/.vscode/settings.json:
```diff
@@ -4,5 +4,5 @@
         "!GetAtt"
     ],
-    "python.pythonPath": "/Users/nhillard/.pyenv/versions/3.8.6/bin/python3.8"
+    "python.pythonPath": "/Users/julialallen/.pyenv/versions/3.8.6/bin/python"
```
I didn't mean to commit this change, but I also lack motivation to undo it.

tadhg-ohiggins replied:
Should the `.vscode/settings.json` file be in the repo at all?

julialeague replied:
Maybe not - the tweaks for debugging and intellisense are useful if you're using vscode, so I kinda like the idea of sharing those. Maybe in a development doc instead?

julialeague replied:
How about this instead? https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/e63356c311077f5321e5da1c367bfac23d46d102

mountHouli replied:
Looks like this `python.pythonPath` line didn't make it into `docs/vs-code.md`.

Do you know what this line does? Should it also be in `docs/vs-code.md`.

I don't know what it does, but I'm wondering if it is related to the info I learned from Nat Hillard and wrote here:
https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/blob/develop/hypertension/docs/design-notes.md#python-environment

mountHouli commented on hypertension/template.yaml:
```diff
@@ -15,6 +15,17 @@ Parameters:
+      - dev
+      - prod
+
+Globals:
```
Cool use of stages.

Are stages required at this point? (I'm just feeling like it might be simplest to avoid dealing with stages until we have a need for them). What do you think?

mountHouli replied:
See my comment on `aws_secrets_manager.py` for more info.

mountHouli commented on hypertension/vro-main/lib/aws_secrets_manager.py:
```diff
@@ -1,8 +1,12 @@
 
 
 def get_lighthouse_rsa_key(awsSecretArn):  # pragma: no cover
+    if os.environ["STAGE"] == "dev":
+        return "development-environment"
+
```
If we are in development, (as soon as my in-progress PR #16  is done) we will still need the lighthouse RSA key--we just need to fetch it from the environment variables of the local machine rather than from AWS Secrets Manager.

I've added support for this in my (in-progress) PR #16  to convert Tadgh's Lighthouse auth script into a python module that the Lambda function will call.

I figure I'll rebase my PR to the end of `develop` after you merge this PR into `develop`, and that way I'll get all your changes in. Then remove the stages and just keep the environment handling code I have in my PR.

julialeague replied:
That sounds good to me - I figured we'd want to add local support for key testing and should have asked you about that. Thanks for being on top of that!

mountHouli commented on .gitignore:
```diff
@@ -1,4 +1,6 @@
 aws-stepfunctions-local-credentials.txt
 .venv
+.metals
```
What's this? :)

julialeague replied:
Great catch! That's an entirely-unneeded Scala server extension that I forgot to disable for our workspace in VSCode (although why I have to disable it in a Python project is a head-scratcher). Updated this to remove the ignore since it's no longer needed by myself 😆 
https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/ff53847da34e1b814befadef604037dd24045b9a#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947L3

mountHouli replied:
(Resolved)

mountHouli commented on hypertension/.vscode/launch.json:
```diff
@@ -1,60 +0,0 @@
-{
```
Thanks for catching this.

When I eliminated Step Functions and switched to having just one Lambda (the VRO Main), I forgot to update this file.

I see that you updated this to work with the VRO Main function, and put it in a suggestion in the `docs/` directory.

I figure that if having the files committed in the `.vscode/` directory doesn't cause problems for anyone, I'm pondering if requiring the manual setup step (rather than just having the files commited in the `.vscode/` directory) has an advantage?

julialeague replied:
I thought about that too, but realized as I looked at the setup that there's still a reference to my local directory in the json that someone using it locally would need to manually update. That's why I ended up going with the doc instead. Let me know what you think about that. I'm happy to change it back if we decided that's the better way.

mountHouli replied:
Got it!

(Resolved)

mountHouli commented on hypertension/Makefile:
```diff
@@ -40,6 +40,10 @@ pytest: build.local
 	$(POETRY) run ptw -c -- --capture=no ; \
 	cd ../..; \
 
+.PHONY: local.sam.invoke
```
Awesome! Thanks for creating a Makefile command for this!

Rename it to `invoke.sam.local` to follow the naming pattern of the others?

julialeague replied:
You got it!
Updated here and in the README: https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/ff53847da34e1b814befadef604037dd24045b9a#diff-deaf427e9209688ccd9e150a750a283b3a7c7b95dc4c73df873689c3b71d0e25L43

mountHouli replied:
(Resolved)

mountHouli commented on hypertension/README.md:
````diff
@@ -23,29 +24,38 @@ Create and fill out our cloud formation template variables
 ```
 
-Note: For deployment to AWS, you'll to get the `KmsCmkId` from the [Initial Deployment](#initial-deployment) section below.
+Note: For deployment to AWS, you'll need to get the `KmsCmkId` from the [Initial Deployment](#initial-deployment) section below.
````
Thanks for catching some typos I made!

julialeague replied:
No problem - we've all made them 😸

mountHouli replied:
(Resolved)

mountHouli commented on hypertension/README.md:
````diff
@@ -23,29 +24,38 @@ Create and fill out our cloud formation template variables
+To build the lambda package locally and run the function in a container:
+
+```sh
+
````
Eliminate blank lines in this code block?

julialeague replied:
You bet! https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/ff53847da34e1b814befadef604037dd24045b9a#diff-deaf427e9209688ccd9e150a750a283b3a7c7b95dc4c73df873689c3b71d0e25L43

mountHouli commented on hypertension/vro-main/layer/fonts/fonts.conf:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
```
What does this do?

julialeague replied:
I'm not sure! it just comes with the `wkhtmltopdf` package.

mountHouli replied:
Do things work without it?

mountHouli commented on hypertension/template.yaml:
```diff
@@ -23,9 +34,13 @@ Resources:
       Runtime: python3.8
       Environment:
         Variables:
+          FONTCONFIG_PATH: /opt/fonts
+          WKHTMLTOPDF_PATH: /opt/bin/wkhtmltopdf
```
How did you pick these directories? (Just curious!)

julialeague replied:
For the FONTCONFIG_PATH, I followed the instructions from wkhtmltopdf's [documentation](https://wkhtmltopdf.org/downloads.html#how-do-i-use-it-in-aws-lambda) for AWS layer deployment. 

For the WKHTMLTOPDF_PATH, that's the default directory in which the binary is installed by Cloudformation; the pdfkit package that depends on wkhtmltopdf can't auto-detect its location, which is what prompted me to set the path as an environment variable.

mountHouli commented on hypertension/vro-main/lib/main.py:
```diff
@@ -1,10 +1,26 @@
     secret = os.environ["LighthousePrivateRsaKeySecretArn"]
     lighthouse_rsa_key = get_lighthouse_rsa_key(secret)
+
+    pdf = get_pdf(event)
```
Make this function take the HTML rather than taking the entire `event` object?

julialeague replied:
Good call: https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/2c503171846b03dc6571e1a9916630fb30968fa1#diff-9b518a5a01017952ebb8a55ba63a59266458aa1cfa3f90fb00519d40d76807beR13-R14

mountHouli commented on hypertension/vro-main/lib/main.py:
```diff
@@ -1,10 +1,26 @@
+
+def get_pdf(event: Dict) -> bytes:
+    pdf_generator = PdfGenerator(WKHTMLTOPDF_PATH)
+    return pdf_generator.generate_from_string(event.get("html"))
```
What do you think of eliminating all the `pdf_generator.py` code and instead doing it in functional style like this:

```py
def get_pdf(event: Dict) -> bytes:
    config = pdfkit.configuration(wkhtmltopdf=WKHTMLTOPDF_PATH)
    return pdfkit.from_string(event.get("html"), False, config)
```

or like this

```py
def get_pdf(html, wkhtmltopdf_path) -> bytes:
    config = pdfkit.configuration(wkhtmltopdf=wkhtmltopdf_path)
    return pdfkit.from_string(html, False, config)
```

?

IMO, it's less lines of code and easier to read. However, I haven't done much OOP in a while, so I could be overlooking something here.

julialeague replied:
I totally agree with you about this. Only difference is that I kept the external file and just took out all the object-y unnecessaries. 
https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/2c503171846b03dc6571e1a9916630fb30968fa1#diff-9b518a5a01017952ebb8a55ba63a59266458aa1cfa3f90fb00519d40d76807beL24

mountHouli commented on hypertension/vro-main/poetry.lock:
```diff
@@ -64,20 +64,20 @@ uvloop = ["uvloop (>=0.15.2)"]
 
 [[package]]
 name = "boto3"
-version = "1.17.90"
```
This is great. Love the idea of keeping packages up to date.

Topic for an in-person VRO engineers discussion:

Do we want to play this by ear (without structure, just doing it) like here, or do we feel we need to add some sort of structure or practice around this. I bet we're fine with no structure for now. Just figure it's worth bringing up.

julialeague replied:
I'm glad you brought that up! I'm used to having a more strict process around library updates. The process was, mainly, we didn't set less-trusted libraries' optimistic locking to minor versions (so that the minor versions wouldn't auto-upgrade to the latest unless we explicitly wanted them to). Also we didn't update the entire lock file when we added or changed one library version. We'd specifically update or install that one library to avoid updating all versions. That likely saved us some production incident headaches, but I don't see the harm in being less strict about it at this point in our project maturity.

Have you had a more strict process around versioning/updates like that? I do think we could revisit this question once we're in some kind of staging environment.

mountHouli commented on hypertension/vro-main/test/conftest.py:
```diff
@@ -16,6 +18,14 @@
     "LighthousePrivateRsaKeySecretArn"
 ] = "Fake. Not used. Doesnt Matter. Just need something. Anything."
 
+# Set the wkhtmltopdf path environment variable by finding the local binary
+wkhtmltopdf_path = subprocess.run(
+    "which wkhtmltopdf", shell=True, capture_output=True
+)
+os.environ["WKHTMLTOPDF_PATH"] = wkhtmltopdf_path.stdout.decode(
+    "utf-8"
+).rstrip("\n")
```
Very cool!

mountHouli commented on hypertension/vro-main/test/main_test.py:
```diff
@@ -1,12 +1,18 @@
+
+
+def test_get_pdf():
+    assert isinstance(get_pdf(test_event), bytes) is True
```
Thanks for adding some tests!

Seems to me it could be difficult to do more than these `isinstance` tests, but I figure I should ask if you thought of anything.

julialeague replied:
I debated adding a dependency to help test this, and @tadhg-ohiggins thought that would be fine, so this is updated to verify the output with pdfminer (and I have it in the other pdf_generator file): https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/2c503171846b03dc6571e1a9916630fb30968fa1#diff-3648e22824e461bf03a66d761604d8f4d3e1db6853c4d736bb2d01f1a1589d99L1

mountHouli reviewed (changes requested):
Great PR! Great write-up!

I had a few feedback/question items!

## Item 1

Let's talk about whether we want the wkhtmltopdf related binaries directly committed to the repo, or whether we want them to have to be built.

## Item 2

Also, in the PR write-up, it looks like the last "card logged" has the same URL as the 2nd one.

## Item 3

I tried to clone a fresh repo and run the unit tests, but it failed. Here's what I did

```sh
aaron-macbook:hypertension houli$ cp cf-template-params-example.env cf-template-params.env

# Then set my env vars in cf-template-params.env

aaron-macbook:hypertension houli$ make local.sam.invoke
cd ./vro-main ; \
	$(which poetry || echo "install poetry. see https://python-poetry.org/") export --without-hashes -f requirements.txt -o requirements.txt --with-credentials ; \
	cd ../..; \

sam build
Building codeuri: /Users/houli/docs/gitrepos/department-of-veterans-affairs/VRO-clean-2/hypertension/vro-main runtime: python3.8 metadata: {} functions: ['VroMainFunction']
Running PythonPipBuilder:ResolveDependencies
Running PythonPipBuilder:CopySource

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided
    
sam local invoke --parameter-overrides $(cat cf-template-params.env) -e vro-main/pdf-event.json
Invoking app.lambda_handler (python3.8)
PdfGeneratorLayer is a local Layer in the template
Image was not found.
Building image.......................................................................................................................................................................................................................................................................................................................................................................................................................................
Skip pulling image and use local one: samcli/lambda:python3.8-abe4fb5f958e42a0ae51cca0a.

Mounting /Users/houli/docs/gitrepos/department-of-veterans-affairs/VRO-clean-2/hypertension/.aws-sam/build/VroMainFunction as /var/task:ro,delegated inside runtime container
START RequestId: 0f8fef53-bf6b-435e-9097-bfb1859bcc89 Version: $LATEST
/var/task/pdfkit/source.py:11: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if self.type is 'file':
Function 'VroMainFunction' timed out after 3 seconds
No response from invoke container for VroMainFunction
aaron-macbook:hypertension houli$ 
```

julialeague commented:
@mountHouli:
> 
> ## Item 1
> Let's talk about whether we want the wkhtmltopdf related binaries directly committed to the repo, or whether we want them to have to be built.

The other valid way to point to the layer content is to provide an arn that points to an S3 bucket object. Since we agreed it's ok to upload the binary to S3, I thought I could update the `template.yml` to point to the arn once it's available to us (once we have it uploaded in VA's internal AWS, I mean). What do you think?
I also think I read that you can publish a lambda layer for reuse whose arn we could reference, but I'm not 100% on that.

> 
> ## Item 2
> Also, in the PR write-up, it looks like the last "card logged" has the same URL as the 2nd one.
> 
Thank you, I fixed that!

> ## Item 3
> I tried to clone a fresh repo and run the unit tests, but it failed. Here's what I did
> 
Ah, I think you ran into the issue I had which led me to create the stage environment variables. It's likely getting stuck on the call to `get_lighthouse_rsa_key`, unless you've set the stage to `dev` in `cf-template-params.env`. Let me know if that works!

julialeague commented:
I made 3 new commits based on our agreement that we should find a way to automate the download and cleanup of the `wkhtmltopdf` binary. I split our layer definitions into their own SAM template, and then wrote python and shell scripts to aid in the deployment of the layers. Those scripts also cover deploying the lambda function, so that we could have a bundled "initial deployment" command to use any time we're deploying the lambda stack into a fresh environment. All of this included updates to our Makefile commands as well. I've tested out the initial deployment script and confirmed that it works as written, given that you have everything in the project set up as documented in our README. I added documentation to the README under the initial deployment section to update the command. I also added a helper script to export temporary AWS sessions, since we have MFA set up in our AWS environments. That script saved me some time after getting tired of manually requesting and exporting the credentials. I think I could still update the README more about making sure your local environment is set up to use MFA properly, so I'll probably add that in another commit, along with whatever else should be changed.

Here are the new commits to review:
https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/083a18823df57ca38d9bfbbb615103765e3a9fea
https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/pull/15/commits/36ea21e7d3b5e499af98560433a16b0f7499a692

mountHouli commented:
Great PR, @julialeague ! Thank you again for figuring out this Layers tech!

Here are the items from the meeting:

## 1: README

Note the initial deployment steps and the subsequent deployment steps.

e.g. Move `make deploy.initial.stack` into the `### Initial Deployment` section. Rename the `### Deployment` section to `### Subsequent Deployments`. In `### Subsequent Deployments`, put the proper steps/commands.

## 2: Standalone Auth and Initial Deployment Cleanup

Move the auth piece into its own script that the user can run from the command line.
- Note: The script should just set the AWS auth env vars (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`) in the current shell, and not mess with the user's `.bashrc`.
- Idea: If you need to get the env vars set, maybe have the README.md say something like
  > To authenticate, run `eval (poetry run python authenticate_to_aws.py)` or something. I dunno.

Moving auth into its own script will allow the auth piece to be taken out of `scripts/initial_deployment/initial_deployment.sh`. This will in turn allow the remainder of that `initial_deployment.sh` content to be moved into the Makefile. (Thank you for explaining this to us!)

End state: `initial_deployment.sh` should disappear.

## 3: Use `poetry run COMMAND` rather than manual activation of the virtual environment

Looks like the cleanup will apply to the content of `initial_deployment.sh` (which will move into the Makefile under item 2), and Makefile `build.initial.deployment.requirements`.

## 4: Poetry + Lambda + Layer Cleanup

Consolidate everything into one poetry project (the one under `vro-main/`)

Make sure the `requirements.txt` ends up _only_ in the Layer that contains the dependencies (and not the Lambda function), so that the Lambda function package size remains small.

Setup different Makefile commands for difference scenarios

1. When just making code changes to the Lambda, without adding a package: Build and deploy just the Lambda
2. When making code changes that include adding a new Python dependency: Build and deploy both the Lambda and the Layer that contains the Python packages.

Update README accordingly.
- Note: This might result in further changes that overwrite some or all of item 1 above.

## 5: Cleanup idempotency comment in `add_deployed_layers_to_template.py`

## Conclusion

Thanks again! I'm stoked about this PR!

julialeague commented:
> ## 1: README
> Note the initial deployment steps and the subsequent deployment steps.
> 
> e.g. Move `make deploy.initial.stack` into the `### Initial Deployment` section. Rename the `### Deployment` section to `### Subsequent Deployments`. In `### Subsequent Deployments`, put the proper steps/commands.
> 

> ## 2: Standalone Auth and Initial Deployment Cleanup
> Move the auth piece into its own script that the user can run from the command line.
> 
> * Note: The script should just set the AWS auth env vars (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`) in the current shell, and not mess with the user's `.bashrc`.
> * Idea: If you need to get the env vars set, maybe have the README.md say something like
>   > To authenticate, run `eval (poetry run python authenticate_to_aws.py)` or something. I dunno.
> 
> Moving auth into its own script will allow the auth piece to be taken out of `scripts/initial_deployment/initial_deployment.sh`. This will in turn allow the remainder of that `initial_deployment.sh` content to be moved into the Makefile. (Thank you for explaining this to us!)
> 
> End state: `initial_deployment.sh` should disappear.
> 
> ## 3: Use `poetry run COMMAND` rather than manual activation of the virtual environment
> Looks like the cleanup will apply to the content of `initial_deployment.sh` (which will move into the Makefile under item 2), and Makefile `build.initial.deployment.requirements`.
> 
> ## 4: Poetry + Lambda + Layer Cleanup
> Consolidate everything into one poetry project (the one under `vro-main/`)
> 
> Make sure the `requirements.txt` ends up _only_ in the Layer that contains the dependencies (and not the Lambda function), so that the Lambda function package size remains small.
> 
>## 5. Setup different Makefile commands for difference scenarios
> 
> 1. When just making code changes to the Lambda, without adding a package: Build and deploy just the Lambda
> 2. When making code changes that include adding a new Python dependency: Build and deploy both the Lambda and the Layer that contains the Python packages.
> 
> Update README accordingly.
> 
> * Note: This might result in further changes that overwrite some or all of item 1 above.
> 
> ## 6: Cleanup idempotency comment in `add_deployed_layers_to_template.py`

- [x] README updates
- [x] Standalone Auth and Initial Deployment Cleanup
- [x] Use poetry run COMMAND
- [x] Poetry + Lambda + Layer Cleanup
- [x] Different Makefile commands for different scenarios
- [x] Cleanup idempotency comment

@mountHouli this is ready for your eyes! I want to note that I found it impossible to move the layers under the current state of the `py-root` directory without them being packaged into the lambda function as well, so I opted to keep the layers directory separate from py-root for now. Let me know if you see any problems with what I did to merge develop into this branch, also!

mountHouli reviewed (changes requested):
Thanks for meeting again today.

Here's a few of the items we talked about.

## 7. Discussion on Consolidating SAM Stacks

(_This item for documentation only. No action necessary._)

### Background

We have our python dependencies in a Lambda layer.

We believe (but are not certain; based on our limited understanding of `sam deploy` change detection) that our development iteration will be faster if we have two workflows:

Case 1: The set of changes is only to the Lambda python code, without adding any python dependencies.

Workflow: Deploy just the lambda.

Case 2: Add a python dependency.

Workflow: Deploy the Lambda + the Layer containing the python dependencies..

We don't know whether SAM CLI will detect whether or not there are changes to the Lambda and/or changes to the Layer, and handle the above two workflows above automatically based on detecting those changes.

Thus, we are leaving the repo is as it stands, with two SAM / CF stacks.

(_I created a Notion ticket in case we want to tackle this in the future._)

## 8. AWS Auth Script Cleanup

- It has its own poetry project. Keep as is.
- Move `stack_deployment/` directly under `hypertension/`.
- Move `export_aws_mfa_credentails.sh` into `stack_deployment/`
- In the README, add a `cd stack_deployment; poetry install` before `source export_aws_mfa_credentials.sh`.
  - Note: Don't miss the `_mfa_` in the name (it's currently missing in the README).
- Change `export_aws_mfa_credentials.sh` to `eval $(poetry run python establish_aws_mfa_session.py)`

## 9. Delete AWS Account Number from Git History

- In `hypertension/template.yaml` re-write the git history to delete at least the AWS account numbers (or the entire Layer ARNs of that's easier) from the commit history.
- Make the Layer ARNs SAM/CF parameters in `hypertension/template.yaml`

julialeague commented:
Closing this PR in favor of #21
